timemachine/src/time_travel.rs

```diff
@@ -0,0 +1,742 @@
+    query_redo_cursor(&self.db)
+  }
+
+  pub fn get_s3_config(
+    &mut self,
+  ) -> Result<Option<DatabaseBackupSourceS3Config>> {
+    let tx = self.db.transaction()?;
+    let s3_bucket = get_config(&tx, TT_CONFIG_KEY_S3_BUCKET)?;
+    let s3_prefix = get_config(&tx, TT_CONFIG_KEY_S3_PREFIX)?;
+
+    let (Some(s3_bucket), Some(s3_prefix)) = (s3_bucket, s3_prefix) else {
+      return Ok(None);
+    };
+
+    Ok(Some(DatabaseBackupSourceS3Config {
+      bucket: s3_bucket,
+      prefix: s3_prefix,
+    }))
+  }
+
+  pub fn init_s3(
+    &mut self,
+    s3_config: &DatabaseBackupSourceS3Config,
+  ) -> Result<()> {
+    let tx = self.db.transaction()?;
+    if get_config(&tx, TT_CONFIG_KEY_S3_INIT_COMPLETED)?.as_deref() == Some("1")
+    {
+      // Validate consistency
+      let current_bucket =
+        get_config(&tx, TT_CONFIG_KEY_S3_BUCKET)?.unwrap_or_default();
+      let current_prefix =
+        get_config(&tx, TT_CONFIG_KEY_S3_PREFIX)?.unwrap_or_default();
+
+      if current_bucket != s3_config.bucket
+        || current_prefix != s3_config.prefix
+      {
+        anyhow::bail!(
+          "Database is previously synced with a different S3 bucket or prefix. If you have moved the backup directory, please use the SQLite CLI to manually delete `s3_init_completed` in table `tt_config`."
+        );
+      }
+
+      return Ok(());
+    }
+    set_config(&tx, TT_CONFIG_KEY_S3_BUCKET, &s3_config.bucket)?;
+    set_config(&tx, TT_CONFIG_KEY_S3_PREFIX, &s3_config.prefix)?;
+    set_config(&tx, TT_CONFIG_KEY_S3_INIT_COMPLETED, "1")?;
+    tx.commit()?;
+    Ok(())
+  }
+
+  pub fn checkout(&mut self, target_versionstamp: [u8; 10]) -> Result<()> {
+    let tx = self.db.transaction()?;
+    if get_config(&tx, TT_CONFIG_KEY_INITIAL_SNAPSHOT_COMPLETED)?.as_deref()
+      != Some("1")
+    {
+      anyhow::bail!("initial snapshot not completed");
+    }
+    let mut target_versionstamp12: [u8; 12] = [0u8; 12];
+    target_versionstamp12[0..10].copy_from_slice(&target_versionstamp[..]);
+    target_versionstamp12[10..12].copy_from_slice(&[0xff, 0xff]);
+    replay_redo_or_undo_in_tx(&tx, target_versionstamp12, true)?;
+    tx.commit()?;
+    Ok(())
+  }
+
+  pub fn get_current_versionstamp(&mut self) -> Result<[u8; 10]> {
+    let tx = self.db.transaction()?;
+    let current_versionstamp12 = hex::decode(
+      get_config(&tx, TT_CONFIG_KEY_CURRENT_SNAPSHOT_VERSIONSTAMP12)?
+        .ok_or_else(|| {
+          anyhow::anyhow!("current_snapshot_versionstamp12 not found")
+        })?,
+    )
+    .ok()
+    .and_then(|x| <[u8; 12]>::try_from(x).ok())
+    .with_context(|| "invalid current_snapshot_versionstamp12")?;
+    assert_eq!(current_versionstamp12[10..12], [0xff, 0xff]);
+    let mut out = [0u8; 10];
+    out.copy_from_slice(&current_versionstamp12[0..10]);
+    Ok(out)
+  }
+
+  pub fn lookup_versionstamps_around_timestamp(
+    &mut self,
+    start: u64,
+    end: u64,
```

Comment:
these u64 are what? Unix timestamps? can you use `DateTime<Utc>` to represent timestamps?

Reply:
yes, updated